src/System.Collections/src/System/Collections/Generic/SortedList.cs

```diff
@@ -161,11 +161,11 @@ public SortedList(IDictionary<TKey, TValue> dictionary, IComparer<TKey> comparer
                 Debug.Assert(count == _keys.Length);
                 if (count > 1)
                 {
-                    Array.Sort<TKey, TValue>(keys, _values, comparer);
-                    comparer = Comparer; // obtain default if this is null.
+                    // Comparer is set to the default if the comparer parameter is null
+                    Array.Sort<TKey, TValue>(keys, _values, Comparer);
                     for (int i = 1; i != keys.Length; ++i)
                     {
-                        if (comparer.Compare(keys[i - 1], keys[i]) == 0)
+                        if (Comparer.Compare(keys[i - 1], keys[i]) == 0)
```

Comment:
`Comparer.Compare` could change `Comparer` so this can't be optimized.

The range checks are probably also not being optimized out. The old and new JIT differ in what range check patterns they can kill. The new JIT is better but now sure if it can work with a non-zero lower bound and a non-`<` comparison.

If you really want to optimize this further you could cache `keys[i - 1]` in a local variable across iterations so that there will not be two heap loads per iteration.